Remove guardt inheritance on exprt [TG-5977]

romainbrenguier opened the pull request:
This is in preparation for providing other implementations of guardt, in particular using BDDs as is done here: https://github.com/diffblue/cbmc/pull/3730

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
Avoid swap by using copy and move

We aim at providing alternative implementations of guardt for which swap
may not be provided.

romainbrenguier committed:
Use 2 copy + 1 move rather than 2 * (copy + swap)

We aim at providing alternative implementations of guardt for which swap
may not be provided.

romainbrenguier committed:
Get rid of unecessary guard declaration

romainbrenguier committed:
Hide inheritance of guardt on exprt

This ensures only methods of the guardt interface are used.

romainbrenguier committed:
Make exprt a field of guard instead of inheriting

Guardt does not override virtual methods or access private fields so we
can remove inheritance and use a field instead.

romainbrenguier committed:
Avoid uneccessary calls to guardt::as_expr()

We aim at providing other implementations of guardt which may not be
encoded using exprt only.
In that case, `as_expr` may not be a trivial operation as it is now, so
it is better to minimize the number of times it gets called.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 344f36824fd4f66fa457e61500c789621a0e22fc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97502712

smowton reviewed (approved):
I leave the relative merits of exprs and BDDs to other reviewers, but I don't see any incorrect refactoring here.

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -430,11 +431,15 @@ bool goto_symex_statet::l2_thread_read_encoding(
-      cond=or_exprt(no_write.op(), cond);
+    {
+      guardt no_write_guard;
+      no_write_guard.add(no_write.op());
```
Please add a constructor that takes an `exprt`.

tautschnig commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -225,7 +225,7 @@ void goto_symext::symex_goto(statet &state)
   if(new_guard.is_true())
   {
-    state.guard = false_exprt();
+    state.guard.add(false_exprt());
```
Could we have something like `state.guard = guardt(false_exprt());` please?

tautschnig reviewed (approved):
Going back to those times where a `guardt` was not an `exprt` does feel a bit weird, but it seems ok. Please add a constructor as indicated below, though.

peterschrammel commented on src/analyses/goto_check.cpp:
```diff
@@ -1405,15 +1405,15 @@ void goto_checkt::add_guarded_claim(
   else
   {
     new_expr=exprt(ID_implies, bool_typet());
+    exprt guard_expr = guard.as_expr();
```
Intermediate local could be removed

peterschrammel commented on src/analyses/goto_rw.h:
```diff
@@ -343,7 +343,7 @@ class rw_guarded_range_set_value_sett:public rw_range_set_value_sett
     const exprt &expr)
   {
-    guard = true_exprt();
+    guard = guardt();
```
For readability, it would be better if there was no default constructor and this takes true expr as argument instead.

romainbrenguier committed:
Remove guardt default constructor

This makes the code clearer when it assumes that the default guard
corresponds to true.

romainbrenguier commented:
@peterschrammel @tautschnig can you check the last commit does what you wanted?

peterschrammel reviewed (approved):
For my part, yes

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -389,8 +389,7 @@ bool goto_symex_statet::l2_thread_read_encoding(
   {
-    guardt write_guard;
-    write_guard.add(false_exprt());
+    guardt write_guard = guardt(false_exprt());
```
I suppose the compiler has to optimise that anyway, but I'd prefer `guardt write_guard(false_exprt());` instead of the assignment operator. (Applies here and below.)

romainbrenguier replied:
I personally prefer `guardt write_guard = guardt(false_exprt());` as `guardt write_guard(false_exprt());` looks a lot like a function declaration.

peterschrammel replied:
I agree with @tautschnig.

tautschnig replied:
@romainbrenguier How about `guardt write_guard{false_exprt{}};` as the coding guidelines suggest anyway?

romainbrenguier replied:
@tautschnig 
> How about guardt write_guard{false_exprt{}}; as the coding guidelines suggest anyway?

Yes I think that's best.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 570502c5ff68f1e00ce481d0f1221ed74524a1e8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97668697
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 3864a0382b20307008e669b73dee61327ca84b6b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97705878
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 69d82ca475d1de84f504fa99544633757186f944).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97951264
